core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/ListTasksResponse.java

```diff
@@ -0,0 +1,154 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.action.admin.cluster.node.tasks.list;
+
+import org.elasticsearch.action.FailedNodeException;
+import org.elasticsearch.action.TaskOperationFailedException;
+import org.elasticsearch.action.support.tasks.BaseTasksResponse;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentBuilderString;
+import org.elasticsearch.common.xcontent.XContentFactory;
+
+import java.io.IOException;
+import java.util.*;
+
+/**
+ * Returns the list of tasks currently running on the nodes
+ */
+public class ListTasksResponse extends BaseTasksResponse implements ToXContent {
+    private static final TaskInfo[] EMPTY = new TaskInfo[0];
+
+    private TaskInfo[] tasks;
+
+    private Map<String, List<TaskInfo>> nodes;
+
+    public ListTasksResponse() {
+    }
+
+    public ListTasksResponse(List<TaskInfo> tasks, List<TaskOperationFailedException> taskFailures, List<? extends FailedNodeException> nodeFailures) {
+        super(taskFailures, nodeFailures);
+        this.tasks = tasks == null ? EMPTY : tasks.toArray(new TaskInfo[tasks.size()]);
+    }
+
+    @Override
+    public void readFrom(StreamInput in) throws IOException {
+        super.readFrom(in);
+        tasks = new TaskInfo[in.readVInt()];
+        for (int i = 0; i < tasks.length; i++) {
+            tasks[i] = new TaskInfo(in);
+        }
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        super.writeTo(out);
+        out.writeVInt(tasks.length);
+        for (TaskInfo task : tasks) {
+            task.writeTo(out);
+        }
+    }
+
+    /**
+     * Returns the list of tasks by node
+     */
+    public Map<String, List<TaskInfo>> getNodes() {
+        if (nodes != null) {
+            return nodes;
+        }
+        Map<String, List<TaskInfo>> nodeTasks = new HashMap<>();
+
+        Set<String> nodes = new HashSet<>();
+        for (TaskInfo shard : tasks) {
+            nodes.add(shard.node());
+        }
+
+        for (String node : nodes) {
+            List<TaskInfo> tasks = new ArrayList<>();
+            for (TaskInfo taskInfo : this.tasks) {
+                if (taskInfo.node().equals(node)) {
+                    tasks.add(taskInfo);
+                }
+            }
+            nodeTasks.put(node, tasks);
+        }
+        this.nodes = nodeTasks;
+        return nodeTasks;
+    }
+
+    public TaskInfo[] getTasks() {
```

Comment:
should we really expose a mutable object? just use the incoming list and copy it?